### ISSUE TITLE
Question about Dynamic Upload Buffers

### ISSUE THREAD

**Issue by Yoshusa** [Microsoft/DirectX-Graphics-Samples#147]
Dear Microsoft Team,

this question is somewhat related to my previous post (#146). Let us assume I have created a buffer resource in the upload heap, and I want to use it as a dynamic resource to upload instance or constant buffer data once per frame. The buffer can be pretty big, e.g. about 1 MB if used for instancing.

Here the questions:
- If I invoke a Draw command, which uses only a section of the (instance) buffer, how much memory is marshalled to the GPU? The whole buffer? The section? The aligned section (e.g. 64 KB)?
- If I use the same unchanged section for another draw call within the same frame, is the buffer then marshalled again (let us assume cache is filled with other data)?

Here some background information about my question: Our engine is divided into render units (e.g. terrain renderer, light renderer, ...). Each render unit has a constant buffer with specific local data. In addition there is one global constant buffer which contains data used by almost all the render units (e.g. current view projection matrix, view position, ...). At the begin of each frame the global constant data is uploaded from the CPU to the upload buffer of the GPU and then read in alot of Draw and Dispatch calls. How often is this global constant buffer marshalled?

Since I'm unsure about the marshalling, our engine currently also creates a global constant buffer in the defauit heap. So at the begin of each frame I copy the data from the CPU to the upload buffer and from there via CopyBufferRegion to the default buffer. So I can be sure performance is well for all the Draw and Dispatch calls using the default buffer. Is this a good practice if dynamic data is read by multiple Draw and Dispatch calls? Or is this overkill?

Kind regards,

Joerg Sahm

**Comment by bobbrow** [comment]
I'm assuming that your reference to "marshalling" comes from the comments in the HelloWorld samples.  I'll start off by saying that I think that comment is misleading, and I'm adding a backlog item to reword it.

Upload heaps reside in system memory and the GPU pulls data into its caches -- the CPU does not push data across the bus to the GPU.  If the GPU cache is invalidated, the GPU will pull the data across the bus from the upload heap in system memory again.  Since this is a more costly operation than reading data that is resident in video memory, our [general guideline](https://msdn.microsoft.com/en-us/library/windows/desktop/dn770374(v=vs.85).aspx) (see the section on UPLOAD heaps) is that upload heaps be used for CPU-write-once, GPU-read-once scenarios.  This isn't a hard rule, since you could possibly get better performance reading more than once from the upload heap on the GPU if the data still resides in the cache.  You'd need to test out your implementation on a variety of hardware with different cache architectures and measure the performance to be sure.

Your last question asks whether copying your upload heap's contents to a default heap is overkill.  If you are reading from the same data multiple times during the frame, I suspect that this is not overkill.  Again, you'd need to profile your app to know for sure.

Let me know if that answers your questions.

**Comment by Yoshusa** [comment]
Hi Bob,

thanks for your reply.

yes, I took the phrase 'marshalling' from the sample, but it was not misleading for me. I've just thought, it was your naming for loading the data by the GPU (I've wrote 'marshalled to the GPU' in my previous post, which is actually pretty bad).

Your reply answered all my questions, but one :)

How much data of a dynamic buffer is pulled across the bus? Let's assume a big dynamic buffer with per instance data, and there is a single instanced draw call using only a small section of the buffer. Does the GPU pull the whole buffer, or just the section needed by the draw call (aligned to the GPU cache block size)? I assume the latter case, but just want to be sure.

Sorry for the question, but at least in the old DirectX 9 days, the drivers sometimes simplified operations by mapping and locking the whole buffer instead of only the requested section.

Cheers,

Joerg
